ndc-models/src/lib.rs

```diff
@@ -420,9 +440,72 @@ pub struct Query {
     pub offset: Option<u32>,
     pub order_by: Option<OrderBy>,
     pub predicate: Option<Expression>,
+    /// Optionally group and aggregate the selected rows
+    pub groups: Option<Grouping>,
 }
 // ANCHOR_END: Query
 
+// ANCHOR: Grouping
+#[skip_serializing_none]
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize, JsonSchema)]
+#[schemars(title = "Grouping")]
+pub struct Grouping {
+    /// Dimensions along which to partition the data
+    pub dimensions: IndexMap<String, Dimension>,
+    /// Aggregates to compute in each group
+    pub aggregates: IndexMap<String, Aggregate>,
+    /// A predicate to apply after grouping rows
+    pub predicate: Option<GroupExpression>,
+}
+// ANCHOR_END: Grouping
+
+// ANCHOR: GroupExpression
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize, JsonSchema)]
+#[serde(tag = "type", rename_all = "snake_case")]
+#[schemars(title = "Group Expression")]
+pub enum GroupExpression {
+    And {
+        expressions: Vec<GroupExpression>,
+    },
+    Or {
+        expressions: Vec<GroupExpression>,
+    },
+    Not {
+        expression: Box<GroupExpression>,
+    },
+    AggregateComparison {
+        aggregate: Aggregate,
+        operator: String,
+        value: AggregateComparisonValue,
+    },
```

Comment:
I think this is now much more similar to the other branch, although I haven't rebased this branch yet, so it's hard to see.

Reply:
It's more similar now, except the difference I raised above about binary and unary operators. What are you thoughts about that?

Reply:
I agree about unary operators (although I think testing for `SUM(..) IS NULL` is a bit weird to be honest, and I can't think of a good use case), but I didn't see anything about binary operators. Did you mean that I should separate out an `AggregateComparisonTarget` with only one variant?

Reply:
I'm not quite sure what you mean, but the changes you've made solve the problem I raised. 🙂

I agree with your assessment of the usefulness of `SUM(..) IS NULL`. I guess I was kinda futureproofing in case there were unary operators that I didn't know about from some weird database that did actually make sense. Personally I don't think it hurts to have the support there, especially since we support them in regular predicates, but I wouldn't object if you wanted to remove unary operator support here, but leave the current structure in place so that it would be easy to incrementally add it back again if we found a real use case for it.